---
title: "Check if Nordic files are into Illumina TOP format"
author: "Paolo Gottardo"
date: "19 7 2016"
---

**Note**
* The genotype files are Nordic format (Illumina long format but without header)
* This script doesn't work on Windows machines. It works on both Linux and Mac OS
* You must have a list of SNPs to check 
* Is not efficient at all....if you have a better alternative use it XD

### Files needed
* File to check. This script works on Illumina long format without headers files......further implementations are needed in order to work with the other Illumina/Affymetrix format files 
* List of the Snps to check. This file is a 2 columns file. The first column contains the name of the SNP and the second column contains the TOP alleles used. You can download the file with this informations from http://bioinformatics.tecnoparco.org/SNPchimp/index.php/download make sure you download the correct type of file

```bash
#!/bin/bash

echo "Give me the name of the file to check"
read -r filename
echo  "Give me the file containing the list of the SNPs to check"
read -r filename1

counter_total=`awk 'END {print NR}' ${filename1}`
counter=0 #counter di fine valutazione
counter_miss=0 #counter snp non trovati



while read line
do
  snpname=$( echo "$line" | cut -f1 )
  echo $snpname
if  fgrep -q  $snpname ${filename}; then #lo snps di culo Ã© in nordic
echo "SNP present in $filename"
if [ $(grep -m 1 "$snpname" ${filename}| cut -f3) == $( grep -m 1 "$snpname" ${filename1}| cut -f2  | cut -d '/' -f1) ] ||  [ $(grep -m 1 "$snpname" ${filename}| cut -f3) == $( grep -m 1 "$snpname" ${filename1}| cut -f2  | cut -d '/' -f2) ] ; then
echo "I'm counting it"
((counter++))
else
echo "ATTENTION Incompatibility found the file could not be in TOP format....."
fi
else
echo "Snp name not found in this file, deleting it from the calculation"
((counter_miss++))
fi
done < ${filename1}
 
 echo " The number of correct counter SNPs is $counter SNPs over $counter_total controlled SNPs "
 echo " The percentage of correct counted SNPs is $((100*counter/(counter_total - counter_miss)))%"
 echo " There are $counter_miss  missing SNPs over $counter_total controlled SNPs "
 echo " The percentage of missing SNPs is  $((100*counter_miss/counter_total))%"
 if [[ $((100*counter/(counter_total - counter_miss))) -ge 100 ]] ; then
 echo " All the SNPs checked and the file is in TOP format"
 elif [[ $((100*counter/(counter_total - counter_miss))) -ge 95 && $((100*counter/(counter_total - counter_miss))) -lt 100 ]] ; then
 echo " There are some discordances.......this suggests that your annotation file contains some mistakes....check it an re-run the script"
 elif [[ $((100*counter/(counter_total - counter_miss))) -lt 95 ]] ; then
 echo " The error rate suggests that your file is not a TOP format Illumina file. Contact your laboratory for more info"
 else
 exit
  fi
```

### Explanation
The script will ask :
* Name of the file to check
* Name of the file containing the list and the type of SNPs to check 

The script will produce :

* Number and percentage of correct checked SNPs 
* Number and percentage of SNPs that are missing in your file according to the SNPs list downloaded 
* A decision if your file is on TOP format or not. This decision is based on a 5% error rate threshold 
